bee-network/src/swarm/protocols/gossip/gossip_io.rs

```diff
@@ -0,0 +1,149 @@
+                    .send(InternalEvent::ProtocolDropped { peer_id })
+                    .expect("The service must not shutdown as long as there are gossip tasks running.");
+
+                break;
+            }
+        }
+
+        // Reasons why this task might end:
+        // (1) The remote dropped the TCP connection.
+        // (2) The local dropped the gossip_in receiver channel.
+
+        debug!("gossip-in: exiting gossip-in processor for {}.", alias!(peer_id));
+    });
+}
+
+async fn recv_valid_message<S>(stream: &mut S, message: &mut [u8], message_len: &mut usize) -> bool
+where
+    S: AsyncRead + Unpin,
+{
+    if let Ok(msg_len) = stream.read(message).await {
+        if msg_len == 0 {
+            false
+        } else {
+            *message_len = msg_len;
+            true
+        }
+    } else {
+        false
+    }
+}
+
+pub fn spawn_gossip_out_processor(
+    peer_id: PeerId,
+    mut writer: WriteHalf<NegotiatedSubstream>,
+    outgoing_gossip_receiver: GossipReceiver,
+    internal_event_sender: InternalEventSender,
+) {
+    tokio::spawn(async move {
+        let mut outgoing_gossip_receiver = outgoing_gossip_receiver.fuse();
+
+        // If the gossip sender dropped we end the connection.
+        while let Some(message) = outgoing_gossip_receiver.next().await {
+            let msg_len = message.len();
+
+            // NB: Instead of polling another shutdown channel, we analogously use an empty message
+            // to signal that we want to end the connection. We use this "trick" whenever the network
+            // receives the `DisconnectPeer` command to enforce that the connection will be dropped.
+
+            if msg_len == 0 {
+                debug!("gossip-out: received shutdown message.");
+
+                // NB: The network service will not shut down before it has received the `ConnectionDropped` event from
+                // all once connected peers, hence if the following send fails, then it must be
+                // considered a bug.
+
+                internal_event_sender
+                    .send(InternalEvent::ProtocolDropped { peer_id })
+                    .expect("The service must not shutdown as long as there are gossip tasks running.");
+
+                break;
+            }
+
+            // If sending to the stream fails we end the connection.
+            if !send_valid_message(&mut writer, &message).await {
+                debug!("gossip-out: stream closed remotely");
+                break;
+            }
+        }
+
+        // Reasons why this task might end:
+        // (1) The local send the shutdown message (len = 0)
+        // (2) The remote dropped the TCP connection.
+
+        debug!("gossip-out: exiting gossip-out processor for {}.", alias!(peer_id));
+    });
+}
+
+async fn send_valid_message<S>(stream: &mut S, message: &[u8]) -> bool
+where
+    S: AsyncWrite + Unpin,
+{
+    if stream.write_all(message).await.is_err() {
```

Comment:
Can be refactored into an if-else chain for clarity

Reply:
can also be one-lined, which I did.